[debops.borgbackup] Add debops.borgbackup role

ypid opened the pull request:
Closes: #792

Do not merge yet, it is not functional currently. Depends on discussion started in: https://github.com/debops/debops/pull/809#issuecomment-500164549

ypid committed:
[debops.borgbackup] Add debops.borgbackup role

ypid committed:
[debops.borgbackup] Actually test the role playbook in CI tests

tumd commented on ansible/roles/debops.borgbackup/templates/etc/ansible/facts.d/mandos.fact.j2:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/python{{ '2' if (ansible_python_version is version_compare('3.5', '<')) else '3' }}
```
See #608
Also is the filename correct? mandos?

ypid committed:
[debops.borgbackup] Fix Ansible facts that where not set before

ypid committed:
[debops.borgbackup] Failed backup run should be auto cleared in mon

.. in Monitoring when it later succeeds.

ypid committed:
[debops.borgbackup] Fix support for Debian buster

TuxCoder commented:
if you like to, I created also a borg backup role for debops, sadly never upstream it, it also get a bit old and would need polishing, you can take a look, probably you find something interesseting.

https://git.o-g.at/ansible/debops.borg_client
https://git.o-g.at/ansible/debops.borg_server

ypid commented:
Thanks @TuxCoder on short peek, only debops.borg_client seems to use `ionice` and `nice` that I have not checked if my role covers this.

This PR needs work which I can hopefully do in the next months since nobody else has jumped in yet.

ypid committed:
[borgbackup] Get the role somewhat up with current DebOps standards

TuxCoder commented on docs/ansible/roles/debops.borgbackup/inventory/group_vars/debops_service_borgbackup.yml:
```diff
@@ -0,0 +1,8 @@
+
+  - name: 'backup-mgt-borgmatic'
+    raw: |
+      backup-mgt ALL = (ALL) NOPASSWD: /usr/local/bin/borgmatic --create --stats
```
I think the path should be dynamic, ether if it is install through package manager or though pip

ypid replied:
Read `SUDOERS(5)` ;)

As you said already, installing via pip is only done up until buster. After that, it will by installed from Debian. If it where not a workaround, `borgbackup__bin_path` could be used.

TuxCoder commented on docs/ansible/roles/debops.borgbackup/inventory/host_vars/backup.example.net/default.yml:
```diff
@@ -0,0 +1,13 @@
+cron__host_jobs:
+
+  'backup':
+    job: '/usr/local/bin/borgmatic --create --stats >> /var/log/backup.log 2>&1'
```
I think the path should be dynamic, ether if it is install through package manager or though pip

ypid replied:
I would like to wait for bullseye to fix this. Does that work for you?

TuxCoder reviewed (commented):
@ypid nice to see that you also implement the virtualenv method for older systems, but I'm not sure why you need it on buster, there is version 1.1.9-2 currently stable, I know you don't wanna use the stretch version, but I think buster should be new enough.
Can you describe your reason for that? IMHO is version 1.1.x new enough with the new encryption method, and other breaking changes.
What kind of feature do you need from the newest version?

Also borgbackup looks over the last 10 years very stable, but I normally don't like to install something from pip that can change on the next deploy, without any notice.

TuxCoder commented:
Another question:
why do you use a different user for each host?

I managed it like this: (user was created by the server role, then different sub folders)
https://git.o-g.at/ansible/debops.borg_client/src/branch/master/tasks/main.yml#L81


Else looks good, I think --append-only is not needed anymore in the newer versions as it has an config entry, but I haven't tested this new config entry yet.
Do you have currently some append only restrictions? Also not sure if Borgmatic does something automatically.

ypid commented:
> why do you use a different user for each host?

[Principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege), [Defense in depth ](https://en.wikipedia.org/wiki/Defense_in_depth_(computing))
> I managed it like this: (user was created by the server role, then different sub folders)

Possible but that still requires on `borg` enforcing this restriction.

> --append-only is not needed anymore in the newer versions as it has an 

Not sure what you mean. I find borgbackup’s `--append-only` interesting obviously, but that is not something that is used in this role yet. But it would allow that clients can only append and a dedicated backup controller does the data retention.

> Also borgbackup looks over the last 10 years very stable, but I normally don't like to install something from pip that can change on the next deploy, without any notice.

I feel the same as you. I also don’t like the idea that on next deployment, something other would happen as defined. This is not something I would accept easily. That is why I hash locked a specific git commit :) Did you overlook that? Note that the thing the role installs with pip is `borgmatic` not `borgbackup`.

> but I'm not sure why you need it on buster, there is version 1.1.9-2 currently stable, I know you don't wanna use the stretch version, but I think buster should be new enough.

As said, the thing the role installs with pip is `borgmatic` not `borgbackup`.

But in case you wonder about borgmatic, which is also in Debian. That version is the first which includes `validate-borgmatic-config` (ref: https://projects.torsion.org/witten/borgmatic/pulls/161), something we want to have.

> Else looks good, I think

This role was basically ready for merging in 2019. Only me insisting on having the configuration format of the role very flexible kept it from getting considered for merging. Note that it still collides with #809. Something I have not thought of when I rebased. So when you test this PR, be sure to revert #809 or better, fix the collision :)

ypid committed:
[borgbackup] Fix borgbackup/borgmatic mix-up that might caused confusion

ypid committed:
chore(borgbackup): Tool assistant use Ansible 2.8 shortcut update

`./bin/debops-optimize`

ypid committed:
chore(borgbackup): Drop old Debian version support

ypid committed:
chore(borgbackup): Align file path defaults

ypid committed:
chore(borgbackup): Manually switch to Ansible 2.8 shortcut

ypid committed:
docs(borgbackup): Fix typos in getting started

ypid committed:
fix(borgbackup): Old backups were never pruned

> If you omit create and other actions, borgmatic runs through a set of
> default actions: prune any old backups as per the configured retention
> policy, compact segments to free up space (with Borg 1.2+), create a
> backup, and check backups for consistency problems due to things like
> file damage

Ref: https://torsion.org/borgmatic/docs/how-to/set-up-backups/

ypid committed:
feature(borgbackup): Update to new redesigned authorized_keys
